cpp/fetcher/continuous_fetcher.cc

```diff
@@ -0,0 +1,157 @@
+using util::Task;
+
+DEFINE_int32(delay_between_fetches_secs, 30, "delay between fetches");
+
+namespace cert_trans {
+
+namespace {
+
+
+class ContinuousFetcherImpl : public ContinuousFetcher {
+ public:
+  ContinuousFetcherImpl(libevent::Base* base, Database<LoggedCertificate>* db);
+
+  void AddPeer(const string& node_id, const shared_ptr<Peer>& peer) override;
+  void RemovePeer(const string& node_id);
+
+ private:
+  void StartFetch(const unique_lock<mutex>& lock);
+  void FetchDone(Task* task);
+  void FetchDelayDone(Task* task);
+
+  libevent::Base* const base_;
+  Database<LoggedCertificate>* const db_;
+
+  mutex lock_;
+  map<string, shared_ptr<Peer>> peers_;
+
+  bool restart_fetch_;
+  unique_ptr<Task> fetch_task_;
+
+  DISALLOW_COPY_AND_ASSIGN(ContinuousFetcherImpl);
+};
+
+
+ContinuousFetcherImpl::ContinuousFetcherImpl(libevent::Base* base,
+                                             Database<LoggedCertificate>* db)
+    : base_(CHECK_NOTNULL(base)),
+      db_(CHECK_NOTNULL(db)),
+      restart_fetch_(false) {
+}
+
+
+void ContinuousFetcherImpl::AddPeer(const string& node_id,
+                                    const shared_ptr<Peer>& peer) {
+  unique_lock<mutex> lock(lock_);
+
+  // TODO(pphaneuf): Allow updating the peer, as this is currently
+  // simplest, for the case where a peer might change host:port.
+  const auto it(peers_.find(node_id));
+  if (it != peers_.end()) {
+    it->second = peer;
+  } else {
+    CHECK(peers_.emplace(node_id, peer).second);
+  }
+
+  if (fetch_task_) {
+    restart_fetch_ = true;
+    fetch_task_->Cancel();
+  } else {
+    StartFetch(lock);
+  }
+}
+
+
+void ContinuousFetcherImpl::RemovePeer(const string& node_id) {
+  lock_guard<mutex> lock(lock_);
+
+  // TODO(pphaneuf): In tests, we rig more than cluster state
+  // controllers to the same continuous fetcher instance, so additions
+  // and removals can be duplicated. Tolerate this for now, but only
+  // restart the fetching process if there was an actual removal.
+  if (peers_.erase(node_id) > 0 && fetch_task_) {
+    restart_fetch_ = true;
+    fetch_task_->Cancel();
+  }
+}
+
+
+void ContinuousFetcherImpl::StartFetch(const unique_lock<mutex>& lock) {
+  CHECK(lock.owns_lock());
+  CHECK(!fetch_task_);
+
+  restart_fetch_ = false;
+
+  unique_ptr<PeerGroup> peer_group(new PeerGroup);
+  for (const auto& peer : peers_) {
+    peer_group->Add(peer.second);
+  }
+
+  fetch_task_.reset(
+      new Task(bind(&ContinuousFetcherImpl::FetchDone, this, _1), base_));
+
+  VLOG(1) << "starting fetch with tree size: " << peer_group->TreeSize();
+  FetchLogEntries(db_, move(peer_group), fetch_task_.get());
+}
+
+
+void ContinuousFetcherImpl::FetchDone(Task* task) {
+  if (!task->status().ok()) {
+    LOG(WARNING) << "error while fetching: " << task->status();
+  }
+
+  lock_guard<mutex> lock(lock_);
+  fetch_task_.reset();
+
+  if (restart_fetch_) {
+    base_->Add(bind(&ContinuousFetcherImpl::FetchDelayDone, this, nullptr));
```

Comment:
does this `nullptr` not just immediately blow up on the `CHECK_NOTNULL` on L137?